cmd/buildah/images.go

```diff
@@ -506,27 +539,37 @@ func formattedSize(size int64) string {
 	return fmt.Sprintf("%.3g %s", formattedSize, suffixes[count])
 }
 
-// reposToMap parses the specified repotags and returns a map with repositories
-// as keys and the corresponding arrays of tags as values.
-func imageReposToMap(repotags []string) map[string][]string {
-	// map format is repo -> tag
+// imageReposToMap parses the specified names list and returns a map with repositories
+// as keys and a slice of tags and digests as values.
+func imageReposToMap(names []string) map[string][]string {
+	// map format is repo -> []tagOrDigest
 	repos := make(map[string][]string)
-	for _, repo := range repotags {
-		var repository, tag string
-		if strings.Contains(repo, ":") {
-			li := strings.LastIndex(repo, ":")
-			repository = repo[0:li]
-			tag = repo[li+1:]
-		} else if len(repo) > 0 {
-			repository = repo
-			tag = "<none>"
-		} else {
+	for _, name := range names {
+		if name == "" {
 			logrus.Warnf("Found image with empty name")
+			continue
+		}
+		named, err := reference.ParseNormalizedNamed(name)
+		if err != nil {
+			logrus.Warnf("Error parsing name %q: %v", name, err)
+			continue
+		}
+		if named.Name() != name {
+			logrus.Debugf("Name %q wasn't already in its normalized form (%q).", name, named.Name())
+		}
+		var repository, tagOrDigest string
+		repository = reference.TrimNamed(named).Name()
+		if tagged, ok := named.(reference.Tagged); ok {
+			tagOrDigest = tagged.Tag()
+		}
+		if digested, ok := named.(reference.Digested); ok {
+			digest := digested.Digest()
+			tagOrDigest = digest.String()
 		}
-		repos[repository] = append(repos[repository], tag)
+		repos[repository] = append(repos[repository], tagOrDigest)
 	}
 	if len(repos) == 0 {
-		repos["<none>"] = []string{"<none>"}
+		repos[""] = []string{""}
```

Comment:
(Non-blocking: This does not _quite_ fit into what this function otherwise does, and notably it muddles the semantic type of the map keys/values: without this the map key is conceptually a `reference.Named`, now it is a “`reference.Named` or empty string”.

This special case exists so that the table output has at least one `<none>:<none>` column, fine, but that is a concern of the caller building the table, and would be a better fit to handle in the caller.)

Reply:
Yup.  I'm trying to avoid reimplementing the caller and the filtering/matching logic, though, because that feels like a much bigger patch.

Reply:
As far as this particular review comment goes, I’d be happy with just moving the `if len(repos) == 0 { repos[""] = []string{""}}` condition to the caller, without changing the logic otherwise.

Anyway, this _is_ non-blocking, and more importantly maybe the map should not exist at all.

Reply:
Easy enough to move that, since this is the only caller.